classes/sso/user/mumie_user_service.php

```diff
@@ -0,0 +1,99 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * This class provides service functionalities for mumie_users
+ *
+ * @package auth_mumie
+ * @copyright  2017-2023 integral-learning GmbH (https://www.integral-learning.de/)
+ * @author Tobias Goltz (tobias.goltz@integral-learning.de)
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace auth_mumie\user;
+
+defined('MOODLE_INTERNAL') || die();
+
+use auth_mumie\hash\hashing_service;
+use auth_mumie\hash\mumie_id_hash;
+
+require_once($CFG->dirroot . '/auth/mumie/classes/sso/hash/hashing_service.php');
+require_once($CFG->dirroot . '/auth/mumie/classes/sso/user/mumie_user.php');
+
+/**
+ * This class provides service functionalities for mumie_users
+ *
+ * @package auth_mumie
+ * @copyright  2017-2023 integral-learning GmbH (https://www.integral-learning.de/)
+ * @author Tobias Goltz (tobias.goltz@integral-learning.de)
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class mumie_user_service {
+    /**
+     * Get a mumie_user instance for a given moodle user and MUMIE Task
+     * @param string    $moodleid
+     * @param \stdClass $mumietask
+     * @return mumie_user
+     */
+    public static function get_user(string $moodleid, \stdClass $mumietask) : mumie_user {
+        if (self::use_id_masking($mumietask)) {
+            $mumieid = hashing_service::generate_hash($moodleid, $mumietask)->get_hash();
+        } else {
+            $mumieid = $moodleid;
+        }
+        return new mumie_user($moodleid, $mumieid);
+    }
+
+    /**
+     * Get mumie_user from a mumie user id.
+     * @param string $mumieid
+     * @return mumie_user
+     * @throws \dml_exception
+     */
+    public static function get_from_mumie_user(string $mumieid) : ?mumie_user {
```

Comment:
I think get_mumie_user is a better function name.

Reply:
We have two similar methods in this class.

`mumie_user_service::get_user` and `mumie_user_service::get_from_mumie_user`. Unfortunately, we cannot overload methods in php.

Both return a `mumie_user` instance. Just naming the method `get_mumie_user` doesn't sound like a better solution for me.

What's the difference between `get_user` and `get_mumie_user`?

I agree that the name should be changed, but I'd suggest something like `get_user_from_mumie_id`.

Reply:
get_user_from_mumie_id is fine

Reply:
done